activation/activation_multi_test.go

```diff
@@ -0,0 +1,483 @@
+			func(ctx context.Context, _ types.NodeID, _ []byte) (*types.Post, *types.PostInfo, error) {
+				<-ctx.Done()
+				return nil, nil, ctx.Err()
+			})
+	}
+
+	require.NoError(t, tab.StartSmeshing(types.Address{}))
+	require.NoError(t, tab.StopSmeshing(true)) // no-op
+
+	for _, sig := range tab.signers {
+		challenge, err := nipost.Challenge(tab.localDb, sig.NodeID())
+		require.ErrorIs(t, err, sql.ErrNotFound)
+		require.Nil(t, challenge) // challenge still deleted
+	}
+}
+
+func TestRegossip(t *testing.T) {
+	layer := types.LayerID(10)
+
+	t.Run("not found", func(t *testing.T) {
+		tab := newTestBuilder(t, 5)
+		for _, sig := range tab.signers {
+			tab.mclock.EXPECT().CurrentLayer().Return(layer)
+			require.NoError(t, tab.Regossip(context.Background(), sig.NodeID()))
+		}
+	})
+
+	t.Run("success", func(t *testing.T) {
+		tab := newTestBuilder(t, 5)
+		var refAtx *types.VerifiedActivationTx
+
+		for _, sig := range tab.signers {
+			atx := newActivationTx(t,
+				sig, 0, types.EmptyATXID, types.EmptyATXID, nil,
+				layer.GetEpoch(), 0, 1, types.Address{}, 1, &types.NIPost{})
+			require.NoError(t, atxs.Add(tab.cdb.Database, atx))
+
+			if refAtx == nil {
+				refAtx = atx
+			}
+		}
+
+		blob, err := atxs.GetBlob(tab.cdb.Database, refAtx.ID().Bytes())
+		require.NoError(t, err)
+
+		// atx will be regossiped once (by the smesher)
+		tab.mclock.EXPECT().CurrentLayer().Return(layer)
+		ctx := context.Background()
+		tab.mpub.EXPECT().Publish(ctx, pubsub.AtxProtocol, blob)
+		require.NoError(t, tab.Regossip(ctx, refAtx.SmesherID))
+	})
+
+	t.Run("checkpointed", func(t *testing.T) {
+		tab := newTestBuilder(t, 5)
+		for _, sig := range tab.signers {
+			require.NoError(t, atxs.AddCheckpointed(tab.cdb.Database,
+				&atxs.CheckpointAtx{ID: types.RandomATXID(), Epoch: layer.GetEpoch(), SmesherID: sig.NodeID()}))
+
+			tab.mclock.EXPECT().CurrentLayer().Return(layer)
+			require.NoError(t, tab.Regossip(context.Background(), sig.NodeID()))
+		}
+	})
+}
+
+func Test_Builder_Multi_InitialPost(t *testing.T) {
+	tab := newTestBuilder(t, 5, WithPoetConfig(PoetConfig{PhaseShift: layerDuration * 4}))
+	var eg errgroup.Group
+	for _, sig := range tab.signers {
+		sig := sig
+		eg.Go(func() error {
+			tab.mnipost.EXPECT().Proof(gomock.Any(), sig.NodeID(), shared.ZeroChallenge).Return(
+				&types.Post{Indices: make([]byte, 10)},
+				&types.PostInfo{
+					CommitmentATX: types.RandomATXID(),
+					Nonce:         new(types.VRFPostIndex),
+				},
+				nil,
+			)
+			tab.mValidator.EXPECT().
+				Post(gomock.Any(), sig.NodeID(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+				AnyTimes().
+				Return(nil)
+
+			require.NoError(t, tab.buildInitialPost(context.Background(), sig.NodeID()))
+
+			// postClient.Proof() should not be called again
+			require.NoError(t, tab.buildInitialPost(context.Background(), sig.NodeID()))
+			return nil
+		})
+	}
+
+	eg.Wait()
+}
+
+func Test_Builder_Multi_HappyPath(t *testing.T) {
```

Comment:
Wouldn't this test be much simpler and more useful as an end-to-end test, with few post-services actually connected? Most of the building blocks are already there in e2e/nipost_test.go. I'm sure that, with the right settings it should be possible to initialize data and run post-services for, say 3 IDs.

Reply:
This test executes in a bit more than a second. The e2e tests take 15+ seconds each and are inherently flaky (sometimes poet is a bit slow or too fast and then the test will fail). I'm pretty sure getting an e2e tests like the ones in the `e2e` package to be stable with multiple identities while not taking a significant amount of time is quite difficult and not needed. I'm planning on adding system tests with multi-smesher nodes in the next PR when everything is integrated between different component and imo this "half step" is more trouble than what it's worth 🙂

Reply:
> Are they? I didn't see CI failing e2e tests recently.

Since I've moved them into their own package I had to adjust them twice for flakiness ^^. I'll see how much effort implementing an e2e like test with multiple identities is.

Reply:
I added an e2e test and fixed a race I found in the PoET client after running the new test with the race detector.